### ISSUE TITLE
kfctl apply fails when using kustomize

### ISSUE THREAD

**Issue by gabrielwen** [kubeflow/kubeflow#3207]
Error message: `the server could not find the requested resource (post gateways.networking.istio.io)`

Witness this at HEAD and v0.5-branch, e.g:
- kfctl built with HEAD and `--version master`
- kfctl built with v0.5-branch and `--version v0.5.0`

**Comment by gabrielwen** [comment]
I feel this is something wrong either from k8s or ISTIO, since we haven't changed related things for quite some time.

Also, I checked with `kubectl get crd` and `gateways.networking.istio.io` is there.

**Comment by lluunn** [comment]
I tested with 0.5.1, basic auth, successful.

**Comment by lluunn** [comment]
nvm, was not using istio, testing again now.

**Comment by lluunn** [comment]
tested with master kfctl, deployed successfully.

**Comment by gabrielwen** [comment]
mm, so it's only me...

**Comment by kkasravi** [comment]
I tried from master and using kustomize which probably isn't valid right now at this time

**Comment by gabrielwen** [comment]
update: I think the scope is on kustomize, tried to deploy from master with ksonnet and it works fine.

**Comment by gabrielwen** [comment]
update: I re-built kfctl from master and the issue regarding ISTIO is gone.

currently the error is: `No Auth Provider found for name "gcp"`

**Comment by gabrielwen** [comment]
tried kubernetes/client-go#242 but the import didn't work.

**Comment by kkasravi** [comment]
@gabrielwen 

TL;DR 
kubeflow/pull/3108 and manifests/pull/58 need to get merged. I've added you to manifests/pull/58 for review

---

this worked for me. Here are the steps

1. KUBEFLOW_USERNAME=kdkasravi KUBEFLOW_PASSWORD=XXX kfctl init ~/dls-kf --platform gcp --project constant-cubist-173123 --use_basic_auth --package-manager kustomize@pull/58 --version pull/3108 --namespace=dls-kf
2. cd dls-kf
3. kfctl generate all --zone us-west1-a
4. KUBEFLOW_USERNAME=kdkasravi KUBEFLOW_PASSWORD=XXX CLIENT_ID=ZZZ CLIENT_SECRET=YYY kfctl apply all -V

**Comment by kkasravi** [comment]
i spoke too soon, i am seeing this errorf

```
INFO[0368] creating tcpkubeattrgenrulerule               filename="app/k8sUtil.go:91"
INFO[0369] creating attributes                           filename="app/k8sUtil.go:91"
INFO[0369] creating istio-policy                         filename="app/k8sUtil.go:91"
INFO[0369] creating istio-telemetry                      filename="app/k8sUtil.go:91"
WARN[0370] Unknown resource: [10 10]                     filename="app/k8sUtil.go:49"
INFO[0371] creating kubeflow-gateway                     filename="app/k8sUtil.go:91"
ERRO[0372] Failed to create kubeflow istio resource: the server could not find the requested resource (post gateways.networking.istio.io)  filename="gcp/gcp.go:719"
Error: couldn't apply KfApp:  (kubeflow.error): Code 500 with message: coordinator Apply failed for gcp:  (kubeflow.error): Code 500 with message: gcp apply could not update deployment manager Error the server could not find the requested resource (post gateways.networking.istio.io)
Usage:
  kfctl apply [all(=default)|k8s|platform] [flags]

Flags:
  -h, --help      help for apply
  -V, --verbose   verbose output default is false
```

I also noticed a number of 'unknown resource' warning in the output

```
INFO[0340] creating istio-sidecar-injector-admin-role-binding-istio-system  filename="app/k8sUtil.go:91"
INFO[0341] creating istio-ingressgateway-sds             filename="app/k8sUtil.go:91"
WARN[0341] Unknown resource: [10 10]                     filename="app/k8sUtil.go:49"
INFO[0341] creating istio-ingressgateway-sds             filename="app/k8sUtil.go:91"
INFO[0341] creating istio-galley                         filename="app/k8sUtil.go:91"
INFO[0342] creating istio-egressgateway                  filename="app/k8sUtil.go:91"
INFO[0342] creating istio-ingressgateway                 filename="app/k8sUtil.go:91"
WARN[0343] Unknown resource: [10 10]                     filename="app/k8sUtil.go:49"
INFO[0343] creating grafana                              filename="app/k8sUtil.go:91"
INFO[0343] creating kiali                                filename="app/k8sUtil.go:91"
INFO[0343] creating istio-policy                         filename="app/k8sUtil.go:91"
INFO[0344] creating istio-telemetry                      filename="app/k8sUtil.go:91"
WARN[0344] Unknown resource: [10 10 10]                  filename="app/k8sUtil.go:49"
INFO[0344] creating istio-pilot                          filename="app/k8sUtil.go:91"
INFO[0345] creating prometheus                           filename="app/k8sUtil.go:91"
INFO[0345] creating istio-citadel                        filename="app/k8sUtil.go:91"
INFO[0346] creating servicegraph                         filename="app/k8sUtil.go:91"
INFO[0346] creating istio-sidecar-injector               filename="app/k8sUtil.go:91"
INFO[0346] creating istio-galley                         filename="app/k8sUtil.go:91"
INFO[0347] creating istio-egressgateway                  filename="app/k8sUtil.go:91"
INFO[0347] creating istio-ingressgateway                 filename="app/k8sUtil.go:91"
WARN[0348] Unknown resource: [10 10]                     filename="app/k8sUtil.go:49"
INFO[0348] creating grafana                              filename="app/k8sUtil.go:91"
INFO[0348] creating kiali                                filename="app/k8sUtil.go:91"
INFO[0349] creating istio-policy                         filename="app/k8sUtil.go:91"
INFO[0349] creating istio-telemetry                      filename="app/k8sUtil.go:91"
INFO[0349] creating istio-pilot                          filename="app/k8sUtil.go:91"
INFO[0350] creating prometheus                           filename="app/k8sUtil.go:91"
INFO[0350] creating istio-citadel                        filename="app/k8sUtil.go:91"
INFO[0351] creating servicegraph                         filename="app/k8sUtil.go:91"
INFO[0351] creating istio-sidecar-injector               filename="app/k8sUtil.go:91"
INFO[0351] creating istio-tracing                        filename="app/k8sUtil.go:91"
INFO[0352] creating istio-egressgateway                  filename="app/k8sUtil.go:91"
INFO[0352] creating istio-ingressgateway                 filename="app/k8sUtil.go:91"
INFO[0353] creating istio-policy                         filename="app/k8sUtil.go:91"
INFO[0353] creating istio-telemetry                      filename="app/k8sUtil.go:91"
INFO[0354] creating istio-pilot                          filename="app/k8sUtil.go:91"
INFO[0354] creating jaeger-query                         filename="app/k8sUtil.go:91"
INFO[0354] creating jaeger-collector                     filename="app/k8sUtil.go:91"
INFO[0355] creating jaeger-agent                         filename="app/k8sUtil.go:91"
INFO[0355] creating zipkin                               filename="app/k8sUtil.go:91"
INFO[0356] creating tracing                              filename="app/k8sUtil.go:91"
INFO[0356] creating istio-sidecar-injector               filename="app/k8sUtil.go:91"
INFO[0357] creating istio-galley                         filename="app/k8sUtil.go:91"
INFO[0357] creating istio-egressgateway                  filename="app/k8sUtil.go:91"
INFO[0357] creating istio-ingressgateway                 filename="app/k8sUtil.go:91"
WARN[0358] Unknown resource: [10 10]                     filename="app/k8sUtil.go:49"
INFO[0358] creating istio-policy                         filename="app/k8sUtil.go:91"
```

**Comment by kkasravi** [comment]
@gabrielwen 

getting close, past installing istio:

```
INFO[0325] creating istio-telemetry                      filename="app/k8sUtil.go:97"
WARN[0325] Unknown resource:
                           filename="app/k8sUtil.go:55"
INFO[0326] creating kubeflow-gateway                     filename="app/k8sUtil.go:97"
INFO[0327] creating grafana-vs                           filename="app/k8sUtil.go:97"
INFO[0327] creating google-api-entry                     filename="app/k8sUtil.go:97"
INFO[0328] creating google-api-vs                        filename="app/k8sUtil.go:97"
INFO[0328] creating google-storage-api-entry             filename="app/k8sUtil.go:97"
INFO[0328] creating google-storage-api-vs                filename="app/k8sUtil.go:97"
INFO[0329] Done installing istio.                        filename="gcp/gcp.go:727"
INFO[0329] Secret for admin-gcp-sa not found, creating ...  filename="gcp/gcp.go:1315"
INFO[0331] Secret for user-gcp-sa not found, creating ...  filename="gcp/gcp.go:1315"
INFO[0332] Secret for admin-gcp-sa not found, creating ...  filename="gcp/gcp.go:1315"
INFO[0333] Secret for user-gcp-sa not found, creating ...  filename="gcp/gcp.go:1315"
WARN[0333] Updating basic auth login failed, trying to create one: secrets "kubeflow-login" not found  filename="gcp/gcp.go:1382"
INFO[0333] reading from /Users/kdkasrav/dls-kf/app.yaml  filename="coordinator/coordinator.go:314"
Error: couldn't apply KfApp:  (kubeflow.error): Code 500 with message: kfApp Apply failed for kustomize@pull/58:  (kubeflow.error): Code 500 with message: could not get core/v1 client
Usage:
  kfctl apply [all(=default)|k8s|platform] [flags]

Flags:
  -h, --help      help for apply
  -V, --verbose   verbose output default is false

couldn't apply KfApp:  (kubeflow.error): Code 500 with message: kfApp Apply failed for kustomize@pull/58:  (kubeflow.error): Code 500 with message: could not get core/v1 client
```

**Comment by gabrielwen** [comment]
/cc @lluunn could you help take a look?

I think for those warning messages we could ignore them - they were also there if we use ksonnet.

**Comment by kkasravi** [comment]
I'm retrying with some potential fixes. I'll update in a bit.

**Comment by kkasravi** [comment]
this now has been fixed but depends on kubeflow/manifests/pull/58 and #3108

**Comment by jlewi** [comment]
@kkasravi @gabrielwen @hougangliu Do we have tests that should catch this?

**Comment by kkasravi** [comment]
Adding @ashahba @swiftdiaries 

We're getting there on tests.

For Generate we have a basic test harness that is based on kustomize's testharness that compares the output of `kustomize build` in a component with the resources/bases for that component. This will be ported to [golang and expanded](kubeflow/manifests/issue/57).

For Apply, I'm working on getting the yaml for the application CR and comparing it with the expected application CR.  This includes health info/status. Since application CRs compose, the tests can also compose with rollup info.

**Comment by jlewi** [comment]
@kkasravi @hougangliu @gabrielwen What about an E2E test like we currently do with ksonnet where we verify that we can actually deploy Kubeflow and then run a variety of checks to verify different applications are running correctly?

**Comment by jlewi** [comment]
It looks like @swiftdiaries has an E2E test in #3060 that has been pending for a while. It looks like the tests might be failing possibly because of this issue?

**Comment by kkasravi** [comment]
@jlewi this bug is waiting on #3108
